### ISSUE TITLE
RAT panics when selected cooperative kernels exceed HIP residency

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#755]
## Summary

On `experimental/RAT`, automatic RAT selection can exhaust its candidate configurations when every selected cooperative kernel requires more resident workgroups than HIP reports. Training then panics instead of selecting a dispatchable configuration or returning a normal error.

## Reproduction

1. Check out `experimental/RAT` at `4aba49d7ee3df1cd1b99765f1a5972a3cb99f759`.
2. Train a composed model that uses `Model::loss(&surrogate)` for one epoch on `amd0`.
3. Run:

```console
recipe --device amd0 example.rs
```

## Actual result

```text
RAT configuration not dispatched: cooperative kernel recipe_model_forward requires 64 resident workgroups; HIP reports 27
RAT configuration not dispatched: cooperative kernel recipe_model_forward requires 40 resident workgroups; HIP reports 27

thread 'main' panicked at recipe.rs:13879:13:
cooperative kernel recipe_model_forward requires 40 resident workgroups; HIP reports 27
```

The process exits with status 101. The failure is reproducible after Recipe finishes loading and compiling the training data.

## Expected result

Automatic RAT selection should continue searching for a configuration whose cooperative workgroup count does not exceed HIP residency. If the search space contains no dispatchable configuration, Recipe should return a regular error that describes the exhausted search space instead of panicking through `Train::run`.

## Environment

- Branch: `experimental/RAT`
- Commit: `4aba49d7ee3df1cd1b99765f1a5972a3cb99f759`
- Device: `amd0`
- Backend: AMD/HIP
- Reported cooperative residency: 27 workgroups

**Comment by nm-z** [comment]
Draft PR #757 now records candidate-dependent residency and shared-memory failures as poor RAT observations. If every explored candidate is rejected, training falls back to the automatic native schedule instead of retrying an invalid candidate. The change remains confined to `recipe.rs`.

**Comment by nm-z** [comment]
The reported panic came from `RatTraining::apply` at 4aba49d7, which returned the residency error once the tuner's initial epochs and shuffles were spent; `Train::execute` then panicked with the same text. On `experimental/RAT` head 73535d36 that return is gone: commit 09032309 (`Complete generic RAT evaluation`, merged through PR #757) records each rejected candidate as `invalid_time_ms`, keeps choosing while the search budget lasts, and when the budget is exhausted with the same candidate still selected it saves the tuner state and dispatches the automatic native schedule (`dispatch(None)`) instead of returning the error.

AMD behaviour was not measured: the native tuner runs only when `rat-enabled` is set and the primary device is `Backend::Amd`, and the only AMD host available could not be used for this check. The evidence here is source-level on 73535d36, not a run on `amd0`. If the automatic schedule itself exceeds HIP residency the run still stops with the residency message, which is the ordinary error path and not the exhausted-search panic this issue reports.
